### PR TITLE
walk: Simplify forward-to logic

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
Currently, forward can be compared with entry names with __XLDIR__
marker, though the code just works, simplify it further for the upcoming
commit.

## Motivation and Context
Code simplification

## How to test this PR?
Listing, listing with continuation

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by klauspost** [comment]
Not really sure I see the simplification - it just looks like the code was moved around.

Please double and triple check that "forward" and "prefix" are completely skipped. If this regresses on that it will have a very negative impact on listing operations.

Since I don't understand the improvement this makes I tend to want to keep the code as-is.

**Comment by vadmeste** [comment]
> Not really sure I see the simplification - it just looks like the code was moved around.

In fact I am trying to send separate PRs to facilitate reviewing. The PR which will come after that will support listing until a given object name and quit early, this is needed to support healing a subset of the namespace (from the beginning until a given object name)

The only confusion in the existing code is that we compare `object__XLDIR__` with 'forward', although it seems it works but this is still making the code hard to read (also it makes hard for the next PR to exist)

This PR:
- Moves the order of code block (should not have an effect)
- Removes __XLDIR__ from the entry name before comparing to 'forward'
- Removes unneeded code at line 251

**Comment by harshavardhana** [comment]
> The only confusion in the existing code is that we compare `object__XLDIR__` with 'forward', although it seems it works but this is still making the code hard to read (also it makes hard for the next PR to exist)

Can you just fix that instead of moving the code around? @vadmeste - let's get that in first and then try to clean things up?

**Comment by klauspost** [comment]
It is quite sensitive wrt ordering. It took several attempts to get the objects ending with `/` ordered correctly, and resume correctly.

So I am a bit hesitant to change too much.

**Comment by vadmeste** [comment]
> It is quite sensitive wrt ordering. It took several attempts to get the objects ending with / ordered correctly, and resume correctly.

> So I am a bit hesitant to change too much.

Yeah I totally understand, I will make more granular changes

File: cmd/metacache-walk.go [hunk 1]
```diff
@@ -194,7 +168,9 @@ func (s *xlStorage) WalkDir(ctx context.Context, opts WalkDirOptions, wr io.Writ
 				meta.name = strings.TrimSuffix(meta.name, SlashSeparator)
 				meta.name = pathJoin(current, meta.name)
 				meta.name = decodeDirObject(meta.name)
-				out <- meta
+				if forward == "" || meta.name >= forward {
```
Comment (klauspost) [inline cmd/metacache-walk.go:171]: It is very expensive that we have to read metadata for every file we are *skipping*.

Before the skip would be done purely based on the name. Now we have to decode and clean up before we skip.

We must be able to skip forward without having to read the file.

**Review by klauspost — CHANGES_REQUESTED** [review]

File: cmd/metacache-walk.go (hunk 1 above)
Comment (vadmeste) [reply cmd/metacache-walk.go:171]: In fact, this is only executed when we are inside an object (it sees xl.meta and part.*). This happens only if you are listing one directory marker

Comment (harshavardhana) [reply cmd/metacache-walk.go:171]: @klauspost do you concur? - do we need this PR @vadmeste - does it optimize reading xl.meta?

Comment (klauspost) [reply cmd/metacache-walk.go:171]: I still don't see the point of the change. Are we fixing anything?

Comment (vadmeste) [reply cmd/metacache-walk.go:171]: > I still don't see the point of the change. Are we fixing anything?

No. My point is to introduce a new listing feature which is called EndAT, which means list until a given object. But this code is hard, that's why I am sending separate PRs to review. It is also hard to do EndAt without this PR (which avoids comparing xl.meta with 'forward'

Comment (harshavardhana) [reply cmd/metacache-walk.go:171]: hmm, interesting can you send the entire PR? @vadmeste hard to imagine what it does :-)

**Comment by harshavardhana** [comment]
@vadmeste are you planning to send a larger change?

**Comment by harshavardhana** [comment]
Going to close this since it looks like you are still working on something locally. Please submit when you have something concrete.

**Pull request closed without merging**